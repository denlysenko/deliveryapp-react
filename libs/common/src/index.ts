export * from './lib/constants';
export * from './lib/errors';
export * from './lib/roles';
export * from './lib/app-menu';
export * from './lib/messages';
export * from './lib/payment-methods';
