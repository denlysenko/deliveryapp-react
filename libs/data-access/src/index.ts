import * as authClient from './lib/api/auth';
import * as usersClient from './lib/api/users';
import * as ordersClient from './lib/api/orders';
import * as paymentsClient from './lib/api/payments';

export { authClient, usersClient, ordersClient, paymentsClient };

export * from './lib/models/auth';
export * from './lib/models/user';
export * from './lib/models/validation-error';
export * from './lib/models/order';
export * from './lib/models/list-response';
export * from './lib/models/orders-filter';
export * from './lib/models/payment';
export * from './lib/models/payments-filter';
export * from './lib/models/users-filter';

export * from './lib/context/auth';
export * from './lib/context/orders';
export * from './lib/context/payments';
