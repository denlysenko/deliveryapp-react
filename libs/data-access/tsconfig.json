{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "jsx": "react",
    "esModuleInterop": true,
    "types": ["node", "jest"]
  },
  "include": ["**/*.ts"]
}
