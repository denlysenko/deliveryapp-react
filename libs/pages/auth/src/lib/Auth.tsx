import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';

import { ACCESS_TOKEN } from '@deliveryapp/common';
import { apiClient } from '@deliveryapp/core';
import {
  authClient,
  AuthCredentials,
  LoginError,
  subscribeToNotifications,
  ValidationError
} from '@deliveryapp/data-access';

import { AuthForm } from './AuthForm/AuthForm';
import { StyledAuth } from './StyledAuth';

export const Auth: React.FC = () => {
  const history = useHistory();

  const [isLoggingIn, setIsLoggingIn] = useState(true);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<LoginError | ValidationError | null>(null);

  const toggleMode = (
    event: React.MouseEvent<HTMLAnchorElement, MouseEvent>
  ) => {
    event.preventDefault();
    setIsLoggingIn((loggingIn) => !loggingIn);
    setError(null);
  };

  const handleSubmit = async (credentials: AuthCredentials) => {
    setLoading(true);

    try {
      const { data } = isLoggingIn
        ? await authClient.login({
            email: credentials.email,
            password: credentials.password
          })
        : await authClient.register(credentials);

      localStorage.setItem(ACCESS_TOKEN, data.token);
      apiClient.setAuthHeader(data.token);
      await subscribeToNotifications();
      history.push('/');
    } catch (err) {
      setLoading(false);
      setError(err.response.data);
    }
  };

  return (
    <StyledAuth>
      <div className="login-panel">
        <div className="p-col-12 logo-container">
          <i className="fa fa-shield"></i>
          <i className="fa fa-sign-in"></i>
          <h2 data-testid="title">
            {isLoggingIn ? 'Login' : 'Register'}
            <a href="/" onClick={toggleMode} data-testid="mode-toggler">
              {isLoggingIn ? 'Register' : 'Login'}
            </a>
          </h2>
        </div>
        <AuthForm
          isLoggingIn={isLoggingIn}
          loading={loading}
          error={error}
          onFormSubmit={handleSubmit}
        />
      </div>
    </StyledAuth>
  );
};
