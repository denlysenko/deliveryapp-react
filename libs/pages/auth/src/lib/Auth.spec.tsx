import React from 'react';
import {
  render,
  cleanup,
  fireEvent,
  wait,
  queryByAttribute,
  Matcher,
  MatcherOptions
} from '@testing-library/react';

import { authPayload, useHistoryMock } from '@deliveryapp/testing';
import { ACCESS_TOKEN } from '@deliveryapp/common';
import { login, register } from '@deliveryapp/data-access';

import { Auth } from './Auth';

type GetByTestId = (
  text: Matcher,
  options?: MatcherOptions | undefined,
  waitForElementOptions?: unknown
) => HTMLElement;

const email = 'test@test.com';
const password = 'password';

const fillFormAndSubmit = (getByTestId: GetByTestId) => {
  fireEvent.change(getByTestId('email'), {
    target: {
      value: email
    }
  });

  fireEvent.change(getByTestId('password'), {
    target: {
      value: password
    }
  });

  fireEvent.click(getByTestId('submit'));
};

jest.mock('react-router-dom', () => ({
  useHistory: () => useHistoryMock
}));

jest.mock('@deliveryapp/data-access', () => ({
  login: jest.fn().mockResolvedValue({ data: authPayload }),
  register: jest.fn().mockResolvedValue({ data: authPayload })
}));

describe('Auth Page', () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  describe('Login', () => {
    it('should render login form', () => {
      const { baseElement } = render(<Auth />);
      expect(baseElement).toMatchSnapshot();
    });

    it('should render correct title', () => {
      const { getByTestId } = render(<Auth />);
      expect(getByTestId('title')).toHaveTextContent('Login');
    });

    it('should render correct toggler', () => {
      const { getByTestId } = render(<Auth />);
      expect(getByTestId('mode-toggler')).toHaveTextContent('Register');
    });
  });

  describe('Register', () => {
    it('should render register form', () => {
      const { baseElement, getByTestId } = render(<Auth />);
      fireEvent.click(getByTestId('mode-toggler'));
      expect(baseElement).toMatchSnapshot();
    });

    it('should render correct title', () => {
      const { getByTestId } = render(<Auth />);
      fireEvent.click(getByTestId('mode-toggler'));
      expect(getByTestId('title')).toHaveTextContent('Register');
    });

    it('should render correct toggler', () => {
      const { getByTestId } = render(<Auth />);
      fireEvent.click(getByTestId('mode-toggler'));
      expect(getByTestId('mode-toggler')).toHaveTextContent('Login');
    });
  });

  describe('Submitting', () => {
    it('should login', async () => {
      const { getByTestId } = render(<Auth />);

      await wait(() => {
        fillFormAndSubmit(getByTestId);
      });

      expect(login).toBeCalledTimes(1);
      expect(login).toBeCalledWith({ email, password });
    });

    it('should register', async () => {
      const { getByTestId } = render(<Auth />);

      fireEvent.click(getByTestId('mode-toggler'));

      await wait(() => {
        fillFormAndSubmit(getByTestId);
      });

      expect(register).toBeCalledTimes(1);
      expect(register).toBeCalledWith({
        email,
        password,
        company: '',
        phone: '',
        firstName: '',
        lastName: ''
      });
    });

    it('should save token to localStorage', async () => {
      const localStorageSpy = jest.spyOn(
        window.localStorage.__proto__,
        'setItem'
      );

      const { getByTestId } = render(<Auth />);

      await wait(() => {
        fillFormAndSubmit(getByTestId);
      });

      expect(localStorageSpy).toBeCalledTimes(1);
      expect(localStorageSpy).toBeCalledWith(ACCESS_TOKEN, authPayload.token);
    });

    it('should redirect to /', async () => {
      const { getByTestId } = render(<Auth />);

      await wait(() => {
        fillFormAndSubmit(getByTestId);
      });

      expect(useHistoryMock.push).toBeCalledTimes(1);
      expect(useHistoryMock.push).toBeCalledWith('/');
    });

    it('should show server errors', async () => {
      const errorMessage = 'Error message';

      (login as jest.MockedFunction<typeof login>).mockImplementationOnce(() =>
        Promise.reject({
          response: {
            data: {
              message: errorMessage
            }
          }
        })
      );

      const { getByTestId, container } = render(<Auth />);

      await wait(() => {
        fillFormAndSubmit(getByTestId);
      });

      expect(
        queryByAttribute('id', container, 'error-message')
      ).toHaveTextContent(errorMessage);
    });
  });
});
