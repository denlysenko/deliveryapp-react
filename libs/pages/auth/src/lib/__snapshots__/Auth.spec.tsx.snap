// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Page Login should render login form 1`] = `
<body>
  <div>
    <div
      class="sc-AykKD AhzHh"
    >
      <div
        class="login-panel ui-fluid"
      >
        <div
          class="ui-g"
        >
          <div
            class="ui-g-12 logo-container"
          >
            <i
              class="fa fa-shield"
            />
            <i
              class="fa fa-sign-in"
            />
            <h2
              data-testid="title"
            >
              Login
              <a
                data-testid="mode-toggler"
                href="/"
              >
                Register
              </a>
            </h2>
          </div>
          <div
            class="sc-AykKC doiYMD"
          >
            <form
              autocomplete="off"
              novalidate=""
            >
              <div
                class="error-message"
                id="error-message"
              >
                <div />
              </div>
              <div
                class="ui-g-12"
              >
                <div
                  class="input-wrapper p-float-label"
                >
                  <input
                    class="p-inputtext p-component"
                    data-testid="email"
                    id="email"
                    value=""
                  />
                  <label
                    for="email"
                  >
                    Email
                  </label>
                  <i
                    class="fa fa-user-circle-o"
                  />
                </div>
              </div>
              <div
                class="ui-g-12"
              >
                <div
                  class="input-wrapper p-float-label"
                >
                  <input
                    class="p-inputtext p-component"
                    data-testid="password"
                    id="password"
                    type="password"
                    value=""
                  />
                  <label
                    for="password"
                  >
                    Password
                  </label>
                  <i
                    class="fa fa-lock"
                  />
                </div>
              </div>
              <div
                class="ui-g-12 button-container"
              >
                <button
                  class="p-button p-component p-button-raised p-button-text-only"
                  data-testid="submit"
                  type="submit"
                >
                  <span
                    class="p-button-text p-c"
                  >
                    Login
                  </span>
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Auth Page Register should render register form 1`] = `
<body>
  <div>
    <div
      class="sc-AykKD AhzHh"
    >
      <div
        class="login-panel ui-fluid"
      >
        <div
          class="ui-g"
        >
          <div
            class="ui-g-12 logo-container"
          >
            <i
              class="fa fa-shield"
            />
            <i
              class="fa fa-sign-in"
            />
            <h2
              data-testid="title"
            >
              Register
              <a
                data-testid="mode-toggler"
                href="/"
              >
                Login
              </a>
            </h2>
          </div>
          <div
            class="sc-AykKC doiYMD"
          >
            <form
              autocomplete="off"
              novalidate=""
            >
              <div
                class="error-message"
                id="error-message"
              >
                <div />
              </div>
              <div
                class="ui-g-12"
              >
                <div
                  class="input-wrapper p-float-label"
                >
                  <input
                    class="p-inputtext p-component"
                    id="firstName"
                    value=""
                  />
                  <label
                    for="firstName"
                  >
                    First Name
                  </label>
                  <i
                    class="fa fa-user-circle-o"
                  />
                </div>
              </div>
              <div
                class="ui-g-12"
              >
                <div
                  class="input-wrapper p-float-label"
                >
                  <input
                    class="p-inputtext p-component"
                    id="lastName"
                    value=""
                  />
                  <label
                    for="lastName"
                  >
                    Last Name
                  </label>
                  <i
                    class="fa fa-user-circle-o"
                  />
                </div>
              </div>
              <div
                class="ui-g-12"
              >
                <div
                  class="input-wrapper p-float-label"
                >
                  <input
                    class="p-inputtext p-component"
                    id="company"
                    value=""
                  />
                  <label
                    for="company"
                  >
                    Company
                  </label>
                  <i
                    class="fa fa-building-o"
                  />
                </div>
              </div>
              <div
                class="ui-g-12"
              >
                <div
                  class="input-wrapper p-float-label"
                >
                  <input
                    class="p-inputtext p-component"
                    id="phone"
                    type="text"
                  />
                  <label
                    for="phone"
                  >
                    Phone
                  </label>
                  <i
                    class="fa fa-phone"
                  />
                </div>
              </div>
              <div
                class="ui-g-12"
              >
                <div
                  class="input-wrapper p-float-label"
                >
                  <input
                    class="p-inputtext p-component"
                    data-testid="email"
                    id="email"
                    value=""
                  />
                  <label
                    for="email"
                  >
                    Email
                  </label>
                  <i
                    class="fa fa-user-circle-o"
                  />
                </div>
              </div>
              <div
                class="ui-g-12"
              >
                <div
                  class="input-wrapper p-float-label"
                >
                  <input
                    class="p-inputtext p-component"
                    data-testid="password"
                    id="password"
                    type="password"
                    value=""
                  />
                  <label
                    for="password"
                  >
                    Password
                  </label>
                  <i
                    class="fa fa-lock"
                  />
                </div>
              </div>
              <div
                class="ui-g-12 button-container"
              >
                <button
                  class="p-button p-component p-button-raised p-button-text-only"
                  data-testid="submit"
                  type="submit"
                >
                  <span
                    class="p-button-text p-c"
                  >
                    Register
                  </span>
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
